src/node.h

```diff
@@ -353,18 +353,81 @@ NODE_EXTERN IsolateData* CreateIsolateData(
 NODE_EXTERN void FreeIsolateData(IsolateData* isolate_data);
 
-// TODO(addaleax): Add an official variant using STL containers, and move
-// per-Environment options parsing here.
+struct ThreadId {
+  uint64_t id = static_cast<uint64_t>(-1);
+};
+NODE_EXTERN ThreadId AllocateEnvironmentThreadId();
+
+namespace EnvironmentFlags {
+enum Flags : uint64_t {
+  kNoFlags = 0,
+  // Use the default behaviour for Node.js instances.
+  kDefaultFlags = 1 << 0,
+  // Controls whether this Environment is allowed to affect per-process state
+  // (e.g. cwd, process title, uid, etc.).
+  // This is set when using kDefaultFlags.
+  kOwnsProcessState = 1 << 1,
+  // Set if this Environment instance is associated with the global inspector
+  // handling code (i.e. listening on SIGUSR1).
+  // This is set when using kDefaultFlags.
+  kOwnsInspector = 1 << 2,
+  // Set up a standard Node.js Environment, including starting any libuv handles
+  // and the inspector, as well as anything dependent on e.g. environment
+  // variables or command line flags.
+  // This conflicts with LoadEnvironment().
+  // It is recommended to not set or rely on this flag, and it will be removed.
+  kPrepareForExecution = 1 << 3,
+  // Override the --abort-on-uncaught-exception CLI flag to never be active
+  // for this Environment instance.
+  kDoNotAbortOnUncaughtException = 1 << 4
+};
+}  // namespace EnvironmentFlags
+
+// TODO(addaleax): Maybe move per-Environment options parsing here.
 // Returns nullptr when the Environment cannot be created e.g. there are
 // pending JavaScript exceptions.
+// It is recommended to use the second variant taking a flags argument.
+// kPrepareForExecution is used for backwards compatibility, but should be
+// removed once LoadEnvironment() supports a more flexible option for
+// embedders than _third_party_main.
 NODE_EXTERN Environment* CreateEnvironment(IsolateData* isolate_data,
                                            v8::Local<v8::Context> context,
                                            int argc,
                                            const char* const* argv,
                                            int exec_argc,
                                            const char* const* exec_argv);
+NODE_EXTERN Environment* CreateEnvironment(
+    IsolateData* isolate_data,
+    v8::Local<v8::Context> context,
+    const std::vector<std::string>& args,
+    const std::vector<std::string>& exec_args,
+    EnvironmentFlags::Flags flags = static_cast<EnvironmentFlags::Flags>(
+        EnvironmentFlags::kDefaultFlags |
+        EnvironmentFlags::kPrepareForExecution),
+    ThreadId thread_id = {} /* allocates a thread id automatically */);
+
+struct InspectorParentHandle {
+  virtual ~InspectorParentHandle();
+};
+NODE_EXTERN std::unique_ptr<InspectorParentHandle> GetInspectorParentHandle(
+    Environment* env,
+    ThreadId thread_id,
+    const char* url);
+
+enum class ProfilerIdleNotifierMode {
```

Comment:
> I am not sure whether we should expose this to the API..this being optional seems rather like an implementation detail?

Is there any downside? In the worst case we can just ignore it, right?

> Would any embedder actually care about this (instead of some higher level setting that suits the use cases as a whole)?

Can you explain what you mean by higher level setting?

But yeah, I don’t know, it’s not all that important I think. And I think we’ve also thought about maybe just always turning this option on in the past.

Reply:
Why happens if an embedder use `kDoNotStart` after passing `--prof` to `ProcessGlobalArgs` and then try to use the CPU profiler (resulting in incorrect profiles?), or use `kStart` without passing `--prof` (no-ops?)? At least for the internal use case it only make sense if you inherit this from the parent environment, then why not just take a parent environment as an argument (races? but how?)?

Reply:
> Whar happens if an embedder uses `kDoNotStart` after passing `--prof` to `ProcessGlobalArgs` and then tries to use the CPU profiler (resulting in incorrect profiles?), or uses `kStart` without passing `--prof` (no-ops?)? At least for the internal use case it only make sense if you inherit this from the parent environment, then why not just take a parent environment as an argument (races? but how?)?

Hm … so, I thought the V8 API CPUProfiler also needs idle notifications, but I’ve looked at `V8CpuProfilerConnection` and it doesn’t appear to be used?

I’m okay with dropping this and always starting the idle notifier for embedder use cases if you prefer. I don’t think we should be making the assumption that there’s always something that could be referred to as a “parent” environment, though.

Reply:
I was referring to this

https://github.com/nodejs/node/blob/0f58bfd0633e357208822a1bc045c5e3bbbb3f6e/src/node.cc#L721-L726

I think only keeping the default behavior (simply using `per_process::v8_is_profiling`) probably makes more sense. I suppose you'd want it to be consistent with whether`--prof` is ever passed to `ProcessGlobalArgs()`, and that's exactly reflected by `per_process::v8_is_profiling`.

Reply:
I’m removing this and switched to always using `per_process::v8_is_profiling` – I guess in that case `InitializeLibuv()` also shouldn’t take an argument?

(Will take a while to push because I’m rebasing instead of pushing fixup commits and I want to make sure everything still compiles/passes after this change)